### DISCUSSION TITLE
Is it appropriate to use this for very large lists?

### DISCUSSION THREAD

**Discussion by noremac** [groue/GRDBQuery#38] [General]
Will this work well for lists that might eventually grow to thousands (or millions?) of rows? The `@Query` demos with GRDB appear to use:

```swift
ValueObservation.tracking { db in try Player.fetchAll(db) }
```

and it's my understanding that would read every player into memory  each time any player changes.

Assuming I'm not missing something, do you have any suggestions/guidance for dealing with larger lists? (I have looked at a few of the demo apps, apologies if I have missed something)

**Comment by groue** [comment]
Hello @noremac,

Dealing with large lists is a know topic of interest, but there's no ready-made api and no demo.

It's partially due to the lack of any "official" way to efficiently deal with large collections in SwiftUI, both in terms of performance and memory consumption. Any time invested on this topic will likely turn obsolete as soon as Apple ships something akin to [Jetpack Paging](https://developer.android.com/topic/libraries/architecture/paging/v3-overview). This is not a great incentive for working on a solution that will need to be complex because "large lists" can mean many things. If you think about it, or just observe your everyday applications, you'll identify a lot of vastly different ways to deal with "large lists".

> and it's my understanding that would read every player into memory each time any player changes.

Your understanding is correct 👍 That's what `fetchAll` does: it returns a plain Swift `Array`, which contains all fetched values.

> do you have any suggestions/guidance for dealing with larger lists?

Yes: a clear idea of what you want, what you want to avoid, and the matching amount of elbow grease, because you'll have to implement it yourself 🙂 

I'm reasonably confident that GRDB has the needed public apis, because the lib has always been about letting the developer do what they need to do, with as little friction as possible, when a feature is not ready-made.

But it is necessary to know what is your goal, and what would be your plan if you had to implement it yourself, even in rough strokes. I could then provide focused guidance.

Oh, and we're dealing with SwiftUI. If you want to display a large list you'll also need to have a good understanding of `List` or `LazyVStack`, including being able to switch to UIKit/AppKit when the UI quality you need requires it - both in terms of performance and memory consumption. So plan for even more elbow grease.
